typescript/ws-protocol/src/FoxgloveServer.ts

```diff
@@ -236,11 +264,47 @@ export default class FoxgloveServer {
 
         break;
 
+      case "advertise":
+        for (const channel of message.channels) {
+          client.advertisements.set(channel.id, channel);
+          this.emitter.emit("advertise", channel);
+        }
+
+        break;
+
+      case "unadvertise":
+        for (const channelId of message.channelIds) {
+          client.advertisements.delete(channelId);
+          this.emitter.emit("unadvertise", channelId);
+        }
+
+        break;
+
       default:
         throw new Error(`Unrecognized client opcode: ${(message as { op: string }).op}`);
     }
   }
 
+  private handleClientBinaryMessage(client: ClientInfo, message: DataView): void {
+    const opcode = message.getUint8(0);
+    switch (opcode) {
+      case BinaryOpcode.MESSAGE_DATA: {
```

Comment:
nit: consider separating client binary opcodes from server binary opcodes? Since the format of the message is not identical.